Tests/Unit/Logger/IoTHubLoggerTests.cs

```diff
@@ -0,0 +1,120 @@
+
+            // act
+            testableLogger.Value.Object.LogInformation(loggedEventId, "Test");
+
+            // assert
+            await VerifySendMessageCountAsync(testableLogger.Value, sendMessageCount);
+        }
+
+        [Fact]
+        public async Task When_Structured_Parameters_Are_Passed_They_Are_Replaced()
+        {
+            // arrange
+            using var testableLogger = SetupProviderAndLogger();
+
+            // act
+            testableLogger.Value.Object.LogInformation("{Id}", 1);
+
+            // assert
+            await VerifyMessageAsync(testableLogger.Value, "1");
+        }
+
+        [Fact]
+        public async void Scope_Is_Set_In_Message_When_Activated()
+        {
+            // arrange
+            using var testableLogger = SetupProviderAndLogger(new IotHubLoggerConfiguration(LogLevel.Trace, default, UseScopes: true));
+            const string devAddr = "12345678ADDR";
+            const string message = "foo";
+
+            // act
+            using var scope = testableLogger.Value.Object.BeginDeviceAddressScope(devAddr);
+            testableLogger.Value.Object.LogInformation(message);
+
+            // assert
+            await VerifyMessageAsync(testableLogger.Value, devAddr + " " + message);
+        }
+
+        [Fact]
+        public async void Scope_Is_Not_Set_In_Message_When_Deactivated()
+        {
+            // arrange
+            using var testableLogger = SetupProviderAndLogger(new IotHubLoggerConfiguration(LogLevel.Trace, default, UseScopes: false));
+            const string devAddr = "12345678ADDR";
+            const string message = "foo";
+
+            // act
+            using var scope = testableLogger.Value.Object.BeginDeviceAddressScope(devAddr);
+            testableLogger.Value.Object.LogInformation(message);
+
+            // assert
+            await VerifyMessageAsync(testableLogger.Value, message);
+        }
+
+        private static DisposableValue<Mock<IotHubLogger>> SetupProviderAndLogger() =>
+            SetupProviderAndLogger(new IotHubLoggerConfiguration(LogLevel.Trace, default, false));
+
+        private static DisposableValue<Mock<IotHubLogger>> SetupProviderAndLogger(IotHubLoggerConfiguration configuration)
+        {
+            var provider = new IotHubLoggerProvider(configuration, new Lazy<Task<ModuleClient>>((Task<ModuleClient>)null!));
+            return new DisposableValue<Mock<IotHubLogger>>(new Mock<IotHubLogger>(provider, null), provider);
+        }
+
+        private static async Task VerifyMessageAsync(Mock<IotHubLogger> logger, string message)
+        {
+#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed (false positive, verification only)
+            await logger.RetryVerifyAsync(x => x.SendAsync(message), Times.Once);
```

Comment:
Here we are testing that our method `SendAsync` was called but not whether the SDK was actually used to send a message. If someone would remove the call to IoTHub these test would still pass. 

Not sure why we need a mock of the whole SUT. Can't we pass a mock only for the "moduleClientFactory" (instead of `null`) and verify calls directly on it?

Reply:
We can't mock the `ModuleClient`, since even though it is non-sealed, it does not have any `virtual` members that we could override/mock. We could discuss if we want to hide external components in a thin wrapper, but that would not solve the problem that the closest layer to the SDK is not unit-testable.